### PR TITLE
bpo-32630: Use contextvars in the decimal module

### PR DESCRIPTION
**Pull request by 1st1**
https://bugs.python.org/issue32630

### CODE REVIEW

File: Modules/_decimal/_decimal.c [hunk 1]
```diff
@@ -122,10 +122,8 @@ incr_false(void)
 }
 
 
-/* Key for thread state dictionary */
-static PyObject *tls_context_key = NULL;
 /* Invariant: NULL or the most recently accessed thread local context */
-static PyDecContextObject *cached_context = NULL;
+static PyContextVar *current_context_var;
```
Comment (njsmith) [inline Modules/_decimal/_decimal.c]: The comment right about this line is wrong now.

File: Modules/_decimal/_decimal.c [hunk 2]
```diff
@@ -1498,117 +1492,65 @@ static PyGetSetDef context_getsets [] =
...
+    Py_DECREF(tok);
 
     /* Borrowed reference with refcount==1 */
     return tl_context;
 }
 
-/* Return borrowed reference to thread local context. */
-static PyObject *
+
+/* Get the context from the thread state dictionary. */
```
Comment (njsmith) [inline Modules/_decimal/_decimal.c]: This doesn't use the thread state dictionary at all anymore?

File: Modules/_decimal/_decimal.c [hunk 3]
```diff
@@ -1498,117 +1492,65 @@ static PyGetSetDef context_getsets [] =
...
+    PyObject *tl_context;
+    if (PyContextVar_Get(current_context_var, NULL, &tl_context) < 0) {
+        return NULL;
+    }
 
-    tstate = PyThreadState_GET();
-    if (cached_context && cached_context->tstate == tstate) {
-        return (PyObject *)cached_context;
+    if (tl_context != NULL) {
+        /* We already have a thread local context. */
```
Comment (njsmith) [inline Modules/_decimal/_decimal.c]: "thread local" is a bit confusing here.

**Review by njsmith — COMMENTED** [review]
Maybe I missed it, but I don't see anything to make `with localcontext(...)` use the token to reset?

**Comment by 1st1** [comment]
> Maybe I missed it, but I don't see anything to make with localcontext(...) use the token to reset?

Correct, I didn't use the `var.reset()` mechanism in this patch to make it as minimal as possible and also make it easier for Stefan to review.

For 3.7 it doesn't really matter if `var.reset()` is used or not since we don't have chained lookups yet.  When we have them, we can just update decimal to use `var.reset()` to be fully compatible.

**Pull request merged**